Add README around how to validate OSCAL (#1170)
usnistgov/OSCAL: README_validations.md

But *validating* goes beyond parsing: it tells you not only that a file can be read successfully, but also that its data structures conform to rules for a particular application of XML, JSON, or YAML. [Well-formed JSON, XML, or YAML is not necessarily valid OSCAL JSON, XML, or YAML](https://pages.nist.gov/OSCAL/concepts/layer/validation/).

If working with XML you will need an XML *schema* (or functional equivalent) to validate your OSCAL. If working with JSON or YAML, you will need a JSON schema. Normative schemas suitable for validating the various OSCAL models are available in the project's source code repository [for XML](https://github.com/usnistgov/OSCAL/tree/main/xml/schema) and [for JSON and YAML](https://github.com/usnistgov/OSCAL/tree/main/json/schema).

## What are the OSCAL models?

OSCAL comes in several different forms, with data models in JSON, XML, and YAML. Please review [this page with the status of all the models as organized into layers](https://pages.nist.gov/OSCAL/concepts/layer/#status-of-the-oscal-layers) for information about each model and the status of its respective schemas.

## OSCAL "complete" schema

> Additionally, there are "complete" XML and JSON schemas, which can validate any kind of OSCAL. It is larger and more complex than the model-specific schemas and may not be as efficient, but it can be convenient to use one schema file as an entrypoint to validate one or more OSCAL documents without knowing the particular model in advance, whether it be a `component-definition`, `system-security-plan`, `assessment-plan`, et cetera.

Reviewer: ```suggestion
Additionally, there are "complete" XML and JSON schemas, which can validate any kind of OSCAL. It is larger and more complex than the model-specific schemas and may not be as efficient, but it can be convenient to use one schema file as an entry point to validate one or more OSCAL documents without knowing the particular model in advance, whether it be a `component-definition`, `system-security-plan`, `assessment-plan`, et cetera.
```

Each tool will have its own way to provide for validation. Bindings or configurations to link documents to schemas can be maintained either inside or outside the documents themselves; many tools offer several ways to do this.

Some tools support self-contained validation functionality without full-fledged application frameworks or development environments, with tools like [`xmllint`](http://xmlsoft.org/xmllint.html). Similarly, you can use the [`ajv-cli`](https://github.com/ajv-validator/ajv-cli) for JSON.

## What does success look like?

If you validate an OSCAL document with a local copy of a schema with `xmllint`, it will return results to the standard output (STDOUT) stream like so:

```
$ xmllint oscal.xml --schema oscal-schema.xsd --noout
oscal.xml validates
```

There is no detailed output, just a statement that indicates validation has been performed successfully.

> If you remove the `--noout` flag from the command line, a copy of the input will be returned. This is because the utility is designed for a pipeline architecture. Its outputs will be provided as inputs to the next tool in the pipeline.

Reviewer: ```suggestion
If you remove the `--noout` flag from the command line, a copy of the input will be returned. This is because the utility is designed for a pipeline architecture, in which its outputs (written to `STDOUT`) can be provided as input to the next tool in the pipeline.
```

  </metadata>
             ^
test-invalid-profile.xml:11: parser error : Extra content at the end of the document
  <merge>
  ^
```

In this case the processor reports the syntax issues it detects, to the extent possible, for a document that is not well-formed. The processor will be unable to complete validation of the document because it is not well-formed.

## What can you learn from errors?

A validation error tells you first and foremost, that your file is incorrect in a basic way. For example, an element (data structure) may be out of place, or missing a required piece. Tools that improperly implement schema validation, misconfigured tools, data quality issues, workflow issues, or many other causes may be the cause. They also vary in how easy they are to correct.

The most important consideration in correcting a validation error is selecting the appropriate schema and the appropriate version of that schema. OSCAL maintains backward compatibility across all schemas as expressed through [the use of semantic versioning](https://github.com/usnistgov/OSCAL/blob/main/docs/content/downloads/_index.md#oscal-releases).

> A *well-formedness error*, however, tells you your file is incorrect in a more fundamental way. It is both more severe (further processing cannot continue) but also easier to fix. In a mature system they should be rare, and when they do occur it is because preliminary checks for well-formedness were not implemented or ignored.

Reviewer: ```suggestion
A *well-formedness error*, however, tells you your file is incorrect in a more fundamental way. It is both more severe (further processing cannot continue) but also easier to address by using tools correctly. In a mature system, well-formedness errors should be rare, and when they do occur it is because preliminary checks for well-formedness were not implemented or ignored.
```

## How can I correct my errors?

End users will ordinarily find it easiest and best to edit a file in an OSCAL tool with a graphical user interface or a structured editor. These would be tools that are especially built or configured to support changing the file while respecting the rulesof well-formedness and validating against the rules of an intended schema.

In order to build such tools and support other operations, developers may also do this work. Commercial off-the-shelf developer environments may support these basic operations for a range of formats.

Finally, XML and JSON are both based on the conventions of *plain-text* (UTF-8 in the case of JSON; any encoding your processor supports in the case of XML). As such, they can be edited in any text editor.

You should validate OSCAL documents saved to files to ensure it will work for the next operation.

## What does validation *not* tell me?

> The rules expressed and enforced by a schema are limited to those that can be systematically checked by automated processes. Valid OSCAL documents are *not necessarily* "correct" from an information perspective. For OSCAL, validation is an objective determination by a program, whereas correct is a subjective determination based on analysis by a person to determine if the information contained in the content is appropriate by some measure. Validation ensures only that *your data is fit for further processing*; such processing can include operations to determine the data's correctness, completeness, adequacy, veracity, or anything else not defined or definable by a schema. 

Reviewer: ```suggestion
The rules expressed and enforced by a schema are limited to those that can be systematically checked by automated processes. Valid OSCAL documents are *not necessarily* "correct" with respect to the information they represent. For OSCAL, validation is an objective determination by a program, whereas "correct" remains a subjective determination based on analysis by a person. Validation ensures only that *your data is fit for further processing*; such processing can include operations to determine the data's correctness, completeness, adequacy, veracity, or anything else not defined or definable by a schema. 
```

Some tools support self-contained validation functionality without full-fledged application frameworks or development environments. Validation against a XML schema in XSD syntax can be done in a single command, for example, with a tool like [`xmllint`](http://xmlsoft.org/xmllint.html) (where `oscal-schema.xsd` is the appropriate schema and `oscal.xml` is an example OSCAL document).

```
$ xmllint oscal.xml --schema oscal-schema.xsd --noout
```

Similarly, using the [`ajv-cli`](https://github.com/ajv-validator/ajv-cli), you can validate a JSON document like so:

```
# Install schema type library in NodeJS runtime to avoid runtime errors
# and run the ajv command line tool after installation.
npm install -g ajv-formats && npx ajv validate -s oscal-schema.json -d oscal.json -c ajv-formats
```

## What does success look like?

If you validate an OSCAL catalog with its schema (a local copy) as above (using `xmllint`), it will return results to the standard output (STDOUT) stream like so:

```
$ xmllint oscal.xml --schema oscal-schema.xsd --noout
oscal.xml validates

> ```

Reviewer 2: Duplicated from above. Why?

Author: That might have been a mistake, checking!

There is no detailed output, just a statement that indicates validation has been performed successfully.

If you remove the `--noout` flag from the command line, a copy of the input will be returned. This is because the utility is designed for a pipeline architecture. Its outputs will be provided as inputs to the next tool in the pipeline.

As for OSCAL JSON, the results from validation with AJV will look like this:

```

> npm install -g ajv-formats && npx ajv validate -s oscal-schema.json -d oscal.json -c ajv-formats

Reviewer 2: Split these commands into two lines. Also, can't ajv be run using `ajv-cli`?

Author: It can be run either with `ajv` or `ajv-cli`, but regardless it will fail without the last bit due to our use of datetimes in schemas. I will change it `ajv-cli`, but `npx` is a helper with modern `npm` so regardless of choice it will install and execute the libraries default execution handler in one command, not two. :-)

[From earlier this week when I forgot this obvious fact with `-c` and why I need to `npm` in the first place.](https://gitter.im/usnistgov-OSCAL/Lobby?at=62263d32257a3578251a9a21).

Author: Also changed locally before pushing back up.

oscal.json valid

> ```

Reviewer 2: Duplicated from above. Why?

## How can I correct my errors?

End users will ordinarily find it easiest and best to edit a file in an OSCAL tool with a graphical user interface or a structured editor. These would be tools that are especially built or configured to support changing the file while respecting the rulesof well-formedness and validating against the rules of an intended schema.

In order to build such tools and support other operations, developers may also do this work. Commercial off-the-shelf developer environments may support these basic operations for a range of formats.

Finally, XML and JSON are both based on the conventions of *plain-text* (UTF-8 in the case of JSON; any encoding your processor supports in the case of XML). As such, they can be edited in any text editor.

You should validate OSCAL documents saved to files to ensure it will work for the next operation.

## What does validation *not* tell me?

> The rules expressed and enforced by a schema are limited to those that can systematically be checked by automated processes. Valid OSCAL documents are *not necessarily* "correct". For OSCAL, validation is an objective determination by program, whereas correct is a subjective determination of analysis by a person. Validataion ensures only that *your data is fit for further processing*; such processing can include operations to determine the data's correctness, completeness, adequacy, veracity, or anything else not defined or definable by a schema. 

Reviewer 2: ```suggestion
The rules expressed and enforced by a schema are limited to those that can be systematically checked by automated processes. Valid OSCAL documents are *not necessarily* "correct" from an information perspective. For OSCAL, validation is an objective determination by a program, whereas correct is a subjective determination based on analysis by a person to determine if the information contained in the content is appropriate by some measure. Validation ensures only that *your data is fit for further processing*; such processing can include operations to determine the data's correctness, completeness, adequacy, veracity, or anything else not defined or definable by a schema. 
```

# How to Validate OSCAL

## XML or JSON?

> XML files use XML syntax: they are *well-formed* by the rules of XML. (See page on wf vs/and valid.) If you can process your file using a conformant XML processor, it is XML.

Reviewer 2: ```suggestion
XML files use XML syntax: they are *well-formed* by the rules of XML. (See page on wf vs/and valid.) If you can process your file using a conformant XML parser, the file is XML.
```

> Same for JSON (although of course you need a JSON processor, not an XML processor).

Reviewer 2: ```suggestion
This same notion applies for JSON and YAML (although of course you need a JSON or YAML parser, instead of an XML parser).
```

> But *validating* goes beyond parsing: it tells you not only that a file is can be read successfully, but also that its data structures conform to rules known ahead or stipulated for a particular application of XML or JSON. All XML is well-formed XML, but only OSCAL XML is valid to the OSCAL models. Likewise with JSON and YAML.

Reviewer 2: ```suggestion
But *validating* goes beyond parsing: it tells you not only that a file can be read successfully, but also that its data structures conform to rules for a particular application of XML, JSON, or YAML. Just because an XML instance is well-formed XML, doesn't mean it is valid OSCAL XML. Likewise with JSON and YAML.
```

If working with XML you will need an XML *schema* (or functional equivalent) to validate your OSCAL. If working with JSON or YAML, you will need a JSON Schema. Normative schemas suitable for validating the various OSCAL models are available on the [project web site]().

> ## What is my model?

Reviewer 2: ```suggestion
## What are the OSCAL models?
```

> OSCAL comes in several different forms, depending on the data set. At time of writing we have the following models. Each will come with its own schema.

Reviewer 2: ```suggestion
OSCAL comes in several different forms, depending on the data set. At the time of writing there are the following OSCAL models. Each comes with its own XML and JSON schema.
```

| Model | XML Schema | JSON Schema |
| -------- | -------- | -------- |
| Catalog | LINK     | LINK     |
| Profile | LINK     | LINK     |

> | System security plan (SSP) | LINK     | LINK     |

Reviewer 2: ```suggestion
| System Security Plan (SSP) | LINK     | LINK     |
```

> | Component definition | LINK     | LINK     |

Reviewer 2: ```suggestion
| Component Definition | LINK     | LINK     |
```

| Assessment Plan | LINK     | LINK     |
| Assessment Results | LINK     | LINK     |
| Plan of Action & Milestones (POAM) | LINK     | LINK     |

Once you have the model and the schema to validate it with, you are ready.

> ## OSCAL "all" schema

Reviewer 2: ```suggestion
## OSCAL "complete" schema
```

> Additionally, for both XML and JSON we have an "all" schema, which can validate any kind of OSCAL, distinguishing them internally. It is larger and more complex than the model-specific schemas and may not be as efficient or useful; but it can be convenient for certain use cases.

Reviewer 2: ```suggestion
Additionally, there are "complete" XML and JSON schemas, which can validate any kind of OSCAL. It is larger and more complex than the model-specific schemas and may not be as efficient or useful, but it can be convenient for certain use cases.
```

## How do I apply a schema to a document to determine validity?

> Schemas are built using standard schema languages, in order that commodity tools can all apply the same schemas to the same documents, and give the same results.

Reviewer 2: ```suggestion
Schemas are built using standard schema languages that are supported by commodity tools. These tools can all apply the same schemas to the same documents, and give the same results.
```

> For OSCAL XML Schemas, this syntax is XSD (XML Schema Definition). For JSON Schemas, it is JSON Schema v7. Free and open-source implementations are available for these; they are also built in to many common application frameworks and dev environments.

Reviewer 2: ```suggestion
For OSCAL XML Schemas, this syntax is XML Schema Definition (XSD). For JSON Schemas, it is JSON Schema. Many freely available and open-source implementations are available for these, as well as commercial tools. These schema standards are also built in to many common application frameworks and development environments.
```

Each tool will have its own way to provide for validation. Bindings or configurations to link documents to schemas can be maintained either inside or outside the documents themselves; many tools offer several ways to do this.

Using a simple command-line tool such as [XML Lint](https://), however, validation against an XML Schema in XSD syntax can be done in a single line.

```
$ xmllint oscal.xml --schema oscal-schema.xsd --noout

```

(where `oscal-schema.xsd` is the appropriate schema.)

Similarly, using AJV, validate a JSON document like this:

```
$ line to validate
```

## What does success look like?

If you validate an OSCAL catalog with its schema (a local copy) as above (using XML Lint), it will return results to STDOUT like so:

```
test-profile.xml validates
```

In other words, no data is returned other than the statement that validation has been performed successfully.

> (If you remove the `--nonout` flag from the command line, a copy of the input will be returned. This is because the utility is designed for a pipeline architecture, whose outputs will be a functio of its inputs.)

Reviewer 2: ```suggestion
If you remove the `--nonout` flag from the command line, a copy of the input will be returned. This is because the utility is designed for a pipeline architecture, whose outputs will be provided as the next tool's inputs.
```
